mlrun/serving/routers.py

```diff
@@ -157,3 +163,256 @@ def _handle_event(self, event):
+            "id": events.body[list(events.body.keys())[0]].body["id"],
+        }
+
+        events.body = event
+        return events
+
+    def do_event(self, event, *args, **kwargs):
+        """handle incoming events, event is nuclio event class"""
+        start = datetime.now()
+        method = event.method or "POST"
+        if event.body and method != "GET":
+            event.body = self.parse_event(event)
+        urlpath = getattr(event, "path", "")
+
+        # if health check or "/" return Ok + metadata
+        if method == "GET" and (
+            urlpath == "/" or urlpath.startswith(self.health_prefix)
+        ):
+            setattr(event, "terminated", True)
+            event.body = self.get_metadata()
+            return event
+
+        # check for legal path prefix
+        if urlpath and not urlpath.startswith(self.url_prefix) and not urlpath == "/":
+            raise ValueError(
+                f"illegal path prefix {urlpath}, must start with {self.url_prefix}"
+            )
+
+        request = self._pre_event_processing_actions(event)
+        response = self.postprocess(self._vote(self._handle_event(event)))
+        if self._model_logger and self.log_router:
+            self._model_logger.push(start, request, response.body)
+        return response
+
+    def _parallel_run(self, event, how="thread"):
+        if how == "array":
+            results = {
+                model_name: model.run(copy.deepcopy(event))
+                for model_name, model in self.routes.items()
+            }
+        else:
+            pool = concurrent.futures.ThreadPoolExecutor(
+                max_workers=len(self.routes) * 2
+            )
+            with pool as executor:
+                results = []
+                futures = {
+                    executor.submit(self.routes[model].run, copy.deepcopy(event)): model
+                    for model in self.routes.keys()
+                }
+                for future in concurrent.futures.as_completed(futures):
+                    child = futures[future]
+                    try:
+                        results.append(future.result())
+                    except Exception as exc:
+                        print("%r generated an exception: %s" % (child.fullname, exc))
+                results = {event.body["model_name"]: event for event in results}
+        return results
+
+    def validate(self, request):
+        """validate the event body (after preprocess)"""
+        if self.protocol == "v2":
+            if "inputs" not in request:
+                raise Exception('Expected key "inputs" in request body')
+
+            if not isinstance(request["inputs"], list):
+                raise Exception('Expected "inputs" to be a list')
+        return request
+
+    def _pre_event_processing_actions(self, event):
+        request = self.preprocess(event.body)
+        if "id" not in request:
+            request["id"] = event.id
+        return self.validate(request)
```

Comment:
the terminology is that `preprocess` is called before anything (such as `parse_event`) is done on the event, see in `BaseModelRouter.do_event` - `preprocess` is called on the first line - can we do the same here ?

Reply:
Changed flow and put request logic in do_event